### PR TITLE
WASM: Generic prefix handling

### PR DESCRIPTION
**Pull request by Cellule**
Implement generic way of dealing with prefixes 
OS#14733708 : Do not allow Tracing opcode from being read by the binary reader.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/4469)

### CODE REVIEW

**Comment by Cellule** [comment]
@arunetm FYI

File: lib/WasmReader/WasmBinaryReader.cpp [hunk 1]
```diff
@@ -438,33 +438,38 @@ const uint32 WasmBinaryReader::EstimateCurrentFunctionBytecodeSize() const
...
 
+WasmOp WasmBinaryReader::ReadPrefixedOpCode(WasmOp prefix, bool isSupported, const char16* notSupportedMsg)
+{
+    CompileAssert(sizeof(WasmOp) >= 2);
+    if (!isSupported)
+    {
+        ThrowDecodingError(notSupportedMsg);
+    }
+    CheckBytesLeft(1);
+    ++m_funcState.count;
```
Comment (Cellule) [inline lib/WasmReader/WasmBinaryReader.cpp]: Right now we don't need to read more than 1 byte past the prefix.
However, I cannot find in the spec nor any proposal if the opcode following the prefix is a byte or a LEB128.
Anyone has more context on this ?

Comment (arunetm) [reply lib/WasmReader/WasmBinaryReader.cpp]: I think the post prefix opcode should remain to be one byte. Trying to confirm this from the spec/discussions.

The design choice of adding the prefix was to restrict the mvp opcodes to a single byte and allow ample space to add post mvp features with significant number of opcodes (like simd). This gave enough space for all post mvp features and cant think of a need for LEB128 opcode encoding (most of the features with prefix are adding much less than 256 new opcodes).

**Comment by arunetm** [comment]
@Cellule fyi
https://github.com/WebAssembly/design/issues/881
https://github.com/WebAssembly/simd/issues/14

File: lib/WasmReader/WasmBinaryReader.cpp (hunk 1 above)
Comment (Cellule) [reply lib/WasmReader/WasmBinaryReader.cpp]: Thanks for the research @arunetm 
Looks like we are doing the right thing regarding prefixes

File: pal/inc/rt/palrt.h [hunk 2]
```diff
@@ -912,7 +912,7 @@ inline int __cdecl _scwprintf_unsafe(const WCHAR *_Format, ...)
 
 inline int __cdecl _vsnwprintf_unsafe(WCHAR *_Dst, size_t _SizeInWords, size_t _Count, const WCHAR *_Format, va_list _ArgList)
 {
-    if (_Count == _TRUNCATE) _Count = _SizeInWords - 1;
+    if (_Count == _TRUNCATE) _Count = _SizeInWords;
```
Comment (Cellule) [inline pal/inc/rt/palrt.h]: @obastemur Can you review the change in PAL ?
I encountered a problem where xplat and Windows builds were producing different result when truncating string with this API

**Comment by Cellule** [comment]
After a few tests failures in xplat builds, I decided to include a few fixes to make xplat correctly run a few more tests in the wasm directory.

**Comment by Cellule** [comment]
ping

File: lib/Common/Memory/HeapAllocator.h [hunk 3]
```diff
@@ -93,6 +93,57 @@ struct HeapAllocatorData
...
+    struct AutoFree
+    {
+    private:
+        T * obj = nullptr;
+    public:
+        ~AutoFree()
+        {
+            HeapDelete(obj);
+        }
+        void Replace(T* obj)
```
Comment (MikeHolman) [inline lib/Common/Memory/HeapAllocator.h]: nit: maybe assert this->obj is null to start/rename this to Set instead of Replace

**Pull request merged**